craft_providers/lxd/lxd_instance.py

```diff
@@ -0,0 +1,494 @@
+
+        :raises subprocess.CalledProcessError: if command fails and check is
+            True.
+        """
+        return self.lxc.exec(
+            instance_name=self.name,
+            command=_rootify_lxc_command(command=command, env=env),
+            project=self.project,
+            remote=self.remote,
+            runner=subprocess.run,
+            **kwargs,
+        )
+
+    def exists(self) -> bool:
+        """Check if instance exists.
+
+        :returns: True if instance exists.
+
+        :raises LXDError: On unexpected error.
+        """
+        return self._get_state() is not None
+
+    def _get_disk_devices(self) -> Dict[str, Any]:
+        """Query instance and return dictionary of disk devices."""
+        devices = self.lxc.config_device_show(
+            instance_name=self.name, project=self.project, remote=self.remote
+        )
+
+        disks = {}
+        for name, config in devices.items():
+            if config.get("type") == "disk":
+                disks[name] = config
+
+                # Ensure the expected keys are in config.
+                if "path" not in config or "source" not in config:
+                    raise LXDError(
+                        brief=f"Failed to parse lxc device {name!r}.",
+                        details=f"* Device configuration: {devices!r}",
+                    )
+
+        return disks
+
+    def _get_state(self) -> Optional[Dict[str, Any]]:
+        """Get state configuration for instance.
+
+        :returns: State information parsed from lxc if instance exists, else
+            None.
+
+        :raises LXDError: On unexpected error.
+        """
+        instances = self.lxc.list(project=self.project, remote=self.remote)
+
+        # lxc returns a filter instances starting with instance name rather
+        # than the exact instance.  Find the exact match...
+        for instance in instances:
+            if instance["name"] == self.name:
+                return instance
+
+        return None
+
+    def is_mounted(self, *, host_source: pathlib.Path, target: pathlib.Path) -> bool:
+        """Check if path is mounted at target.
+
+        :param host_source: Host path to check.
+        :param target: Instance path to check.
+
+        :returns: True if host_source is mounted at target.
+
+        :raises LXDError: On unexpected error.
+        """
+        disks = self._get_disk_devices()
+
+        return any(
+            disk["path"] == target.as_posix()
+            and disk["source"] == host_source.as_posix()
+            for _, disk in disks.items()
+        )
+
+    def is_running(self) -> bool:
+        """Check if instance is running.
+
+        :returns: True if instance is running.
+
+        :raises LXDError: On unexpected error.
+        """
+        state = self._get_state()
+        if state is None:
+            raise LXDError(brief=f"Instance {self.name!r} does not exist.")
+
+        return state.get("status") == "Running"
```

Comment:
Can this be affected by locale?

Reply:
Good question, should set it here. will fix!

Reply:
good catch

Reply:
It does not look like lxc has localization.

Reply:
Not now, but it may have in the future. Setting `LC_ALL` to something sane such as `C` or `POSIX` before collecting this data is a safer approach, but we can do this later.